server/sonar-process-monitor/src/test/java/org/sonar/application/SchedulerImplTest.java

```diff
@@ -0,0 +1,443 @@
+    // leader takes the lock, so underTest won't get it
+    assertThat(appState.tryToLockWebLeader()).isTrue();
+
+    appState.setOperational(ProcessId.ELASTICSEARCH);
+    enableAllProcesses();
+    SchedulerImpl underTest = newScheduler();
+    underTest.schedule();
+
+    processLauncher.waitForProcessAlive(ProcessId.ELASTICSEARCH);
+    assertThat(processLauncher.processes).hasSize(1);
+
+    // leader becomes operational -> follower can start
+    appState.setOperational(ProcessId.WEB_SERVER);
+    processLauncher.waitForProcessAlive(WEB_SERVER);
+
+    underTest.terminate();
+  }
+
+  @Test
+  public void web_server_waits_for_remote_elasticsearch_to_be_started_if_local_es_is_disabled() throws Exception {
+    settings.set(ProcessProperties.CLUSTER_ENABLED, "true");
+    settings.set(ProcessProperties.CLUSTER_SEARCH_DISABLED, "true");
+    SchedulerImpl underTest = newScheduler();
+    underTest.schedule();
+
+    // WEB and CE wait for ES to be up
+    assertThat(processLauncher.processes).isEmpty();
+
+    // ES becomes operational on another node -> web leader can start
+    appState.setRemoteOperational(ProcessId.ELASTICSEARCH);
+    processLauncher.waitForProcessAlive(WEB_SERVER);
+    assertThat(processLauncher.processes).hasSize(1);
+
+    underTest.terminate();
+  }
+
+  @Test
+  public void compute_engine_waits_for_remote_elasticsearch_and_web_leader_to_be_started_if_local_es_is_disabled() throws Exception {
+    settings.set(ProcessProperties.CLUSTER_ENABLED, "true");
+    settings.set(ProcessProperties.CLUSTER_SEARCH_DISABLED, "true");
+    settings.set(ProcessProperties.CLUSTER_WEB_DISABLED, "true");
+    SchedulerImpl underTest = newScheduler();
+    underTest.schedule();
+
+    // CE waits for ES and WEB leader to be up
+    assertThat(processLauncher.processes).isEmpty();
+
+    // ES and WEB leader become operational on another nodes -> CE can start
+    appState.setRemoteOperational(ProcessId.ELASTICSEARCH);
+    appState.setRemoteOperational(ProcessId.WEB_SERVER);
+
+    processLauncher.waitForProcessAlive(COMPUTE_ENGINE);
+    assertThat(processLauncher.processes).hasSize(1);
+
+    underTest.terminate();
+  }
+
+  private SchedulerImpl newScheduler() {
+    return new SchedulerImpl(settings, appReloader, javaCommandFactory, processLauncher, appState)
+      .setProcessWatcherDelayMs(1L);
+  }
+
+  private Scheduler startAll() throws InterruptedException {
+    enableAllProcesses();
+    SchedulerImpl scheduler = newScheduler();
+    scheduler.schedule();
+    processLauncher.waitForProcess(ELASTICSEARCH).operational = true;
+    processLauncher.waitForProcess(WEB_SERVER).operational = true;
+    processLauncher.waitForProcess(COMPUTE_ENGINE).operational = true;
+    return scheduler;
+  }
+
+  private void waitForAppStateOperational(ProcessId id) throws InterruptedException {
+    while (true) {
+      if (appState.isOperational(id, true)) {
+        return;
+      }
+      Thread.sleep(1L);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)